src/hil/spi_master.rs

```diff
@@ -1,73 +1,86 @@
-//! Traits and parameters for SPI master communication
+//! Traits and parameters for SPI master communication.
+//! Authors: Sam Crow <samcrow@uw.edu>
+//!          Philip Levis <pal@cs.stanfored.edu>
 
 use core::option::Option;
 
-#[derive(Copy, Clone)]
-pub enum Rate {
-    MSBFirst,
-    LSBFirst,
-}
 /// Values for the ordering of bits
 #[derive(Copy, Clone)]
-pub enum DataOrder {
-    MSBFirst,
-    LSBFirst,
-}
+pub enum DataOrder {MSBFirst, LSBFirst}
 
 /// Values for the clock polarity (idle state or CPOL)
 #[derive(Copy, Clone)]
-pub enum ClockPolarity {
-    IdleHigh,
-    IdleLow,
-}
-/// Values for the clock phase (CPHA), which defines when
-/// values are sampled
+pub enum ClockPolarity {IdleHigh, IdleLow}
+
+/// Which clock edge values are sampled on
 #[derive(Copy, Clone)]
-pub enum ClockPhase {
-    SampleLeading,
-    SampleTrailing,
-}
+pub enum ClockPhase {SampleLeading, SampleTrailing}
 
 pub trait SpiCallback {
     /// Called when a read/write operation finishes
-    fn read_write_done(&'static self, 
-                       read: Option<&'static mut[u8]>,
-                       writer: Option<&'static mut[u8]>);
```

Comment:
The application writes with app slice A. The driver copies A into an internal kernel buffer B, and invokes the HAL with B.

The application reads into app slice A. The HAL invokes the HAL with internal kernel buffer B, and when the read completes copies it into A.

Right?